Add ability to specify ranges for random numbers in boundary corrections test helper

fmahebert opened the pull request:
## Proposed changes

Depends on #2967 

Primary change:
- enable calls to the generic boundary correction test functions to specify the [min,max) interval from which to generate random numbers for each Tag. By default, any Tag will be drawn from [-1,1), but the new feature enables specifying an arbitrary interval.

Also:
- various smaller cleanups to increase legibility and/or consistency with boundary conditions test helper

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Add comment to bdry condition test helper

fmahebert committed:
Rename function in bdry correction test helper

fmahebert committed:
Add ranges to bdry correction test helper

fmahebert committed:
Move code line to declare variable just before use

fmahebert committed:
Reorder code blocks to match across bdry corr, cond test helpers

fmahebert committed:
Split Tags::Range from bdry condition test helper

nilsdeppe commented on tests/Unit/Evolution/Systems/NewtonianEuler/BoundaryCorrections/Test_Rusanov.cpp:
```diff
@@ -76,19 +77,27 @@ struct DummyInitialData {
+      helpers::Tags::Range<NewtonianEuler::Tags::MassDensityCons>,
+      helpers::Tags::Range<
+          NewtonianEuler::Tags::SpecificInternalEnergy<DataVector>>>
+      ranges{std::array<double, 2>{{1e-6, 1.0}},
```
I'm genuinely unsure of how small the lower bound here should be. Since ideally our evolution scheme should support down to something like `1e-30` should we lower to that?

Also `1e-6` -> `1.0e-6`

fmahebert replied:
I tried this out with the range (1e-30, 1e-29) to make sure I'd get something in the "very small" end of the range. Things seem ok. I'm squashing in a change of the lower bound to 1e-30. Squashed in the notation change too.

nilsdeppe commented on tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp:
```diff
@@ -213,6 +201,10 @@ void test_boundary_correction_conservation_impl(
                          curved_background>::template f<System>>>,
       face_tags>;
 
+  MAKE_GENERATOR(gen);
```
Not a change request since it's beyond the scope of this PR, but what I've learned over time is that the generators are ideally always passed in. The reason being that this makes specifying a seed much easier: you look in the `SPECTRE_TEST_CASE` for the `MAKE_GENERATOR` call and specify the seed. I think changing the `_impl` to receive the `gen` should be not too tedious, while changing the interface to take the generators, would be more annoying.

fmahebert replied:
I almost did this as part of this PR before submitting it... since you feel this way too I've added a commit. :)

nilsdeppe replied:
Haha noice :D Well thank you!

nilsdeppe reviewed (changes requested):
LGTM, one small request and some musings. Feel free to squash anything you decide to change. Thank you for doing this :)

fmahebert committed:
Update bdry correction test helper to take generator as argument

fmahebert commented:
Rebased on develop now that no longer dependent.

Per Nils' review, I've also
- squashed in the small bounds changes
- added a new commit changing the test helper interface to take in the `generator`
